plugins/inmemoryscanner/src/lib/Scanner.cpp

```diff
@@ -45,76 +48,105 @@ namespace InMemoryScanner
 
     bool Scanner::shouldRegionBeScanned(const MemoryRegion& memoryRegionDescriptor)
     {
-        bool verdict = true;
         if (configuration->isScanAllRegionsActivated())
         {
             return true;
         }
         if (memoryRegionDescriptor.isSharedMemory && !memoryRegionDescriptor.isProcessBaseImage)
         {
-            verdict = false;
             logger->info("Skipping: Is shared memory and not the process base image.");
+            return false;
         }
-        return verdict;
+        return true;
     }
 
-    void
-    Scanner::scanMemoryRegion(pid_t pid, const std::string& processName, const MemoryRegion& memoryRegionDescriptor)
+    std::vector<uint8_t> Scanner::constructPaddedMemoryRegion(std::span<const MappedRegion> regions)
+    {
+        std::vector<uint8_t> result;
+
+        if (regions.empty())
+        {
+            return result;
+        }
+
+        std::size_t regionSize = 0;
+        for (const auto& region : regions)
+        {
+            regionSize += region.asSpan().size();
+            regionSize += pageSizeInBytes;
+        }
+        // last region should not have succeeding padding page
+        regionSize -= pageSizeInBytes;
+
+        result.reserve(regionSize);
+        // copy first region
+        auto frontRegionSpan = regions.front().asSpan();
+        std::ranges::copy(frontRegionSpan.begin(), frontRegionSpan.end(), std::back_inserter(result));
+
+        for (std::size_t i = 1; i < regions.size(); i++)
+        {
+            const auto& region = regions[i];
+            // padding page
+            result.insert(result.end(), pageSizeInBytes, 0);
+            auto regionSpan = region.asSpan();
+            std::ranges::copy(regionSpan.begin(), regionSpan.end(), std::back_inserter(result));
+        }
+
+        return result;
+    }
+
+    void Scanner::scanMemoryRegion(pid_t pid,
+                                   addr_t dtb,
+                                   const std::string& processName,
+                                   const MemoryRegion& memoryRegionDescriptor)
     {
         logger->info("Scanning Memory region",
                      {{"VA", fmt::format("{:x}", memoryRegionDescriptor.base)},
                       {"Size", memoryRegionDescriptor.size},
                       {"Module", memoryRegionDescriptor.moduleName}});
 
-        if (shouldRegionBeScanned(memoryRegionDescriptor))
+        if (!shouldRegionBeScanned(memoryRegionDescriptor))
         {
-            auto scanSize = memoryRegionDescriptor.size;
-            auto maximumScanSize = configuration->getMaximumScanSize();
-            if (scanSize > maximumScanSize)
-            {
-                logger->info("Memory region is too big, reducing to maximum scan size",
-                             {{"Size", scanSize}, {"MaximumScanSize", maximumScanSize}});
-                scanSize = maximumScanSize;
-            }
+            return;
```

Comment:
It is logged in `shouldRegionBeScanned`. I also thought that it would be counterintuitive, but I've kept it that way nonetheless. Want me to change it?

Reply:
Looking at shouldRegionBeScanned it is not so clear anymore. I mean the logging tells us exactly why the function returns false. But if we move the exact logging to scanMemoryRegion we can not be sure that it is still valid (because there could be other reasons why shouldRegionBeScanned returns false). On the other hand if we just say "skipping this region" we loose information in the logs.

Just use whatever approach you deem best.